Issue #167: curl step

marcghaly opened the pull request:
Hi @brendanheywood,

This is PR for issue #167,

Regards,

Marc

brendanheywood commented on classes/local/step/connector_curl.php:
```diff
@@ -26,22 +26,183 @@
  */
 class connector_curl extends connector_step {
+    /** @var int[] number of input flows (min, max). */
+    protected $inputflows = [0, 0];
```
It is valid to have a flow step which finishes up, and then a connector step after it. The defaults have been fixed in the connector_step base so you can remove most of these

marcghaly replied:
Fixed.

keevan replied:
Seems like these are still here. I believe you can remove both $inputflows and $inputconnectors properties from this class.

brendanheywood commented on lang/en/tool_dataflows.php:
```diff
@@ -229,3 +229,13 @@
 $string['trigger_cron:next_run_time'] = 'Next run time: {$a}';
+
+// Connector cURL.
+$string['connector_curl:curl'] = 'Client URL';
```
need a version.php bump to pick these up

keevan replied:
You'll need to rebump the version again since the changes might have been lost in your rebase.

marcghaly replied:
Fixed.

brendanheywood commented on classes/local/step/connector_curl.php:
```diff
@@ -26,22 +26,183 @@
+     * @param \MoodleQuickForm &$mform
+     */
+    public function form_add_custom_inputs(\MoodleQuickForm &$mform) {
+        $mform->addElement('textarea', 'config_curl', get_string('connector_curl:curl', 'tool_dataflows'),
```
This should be a single line not a text area

brendanheywood commented on classes/local/step/connector_curl.php:
```diff
@@ -26,22 +26,183 @@
+    public function form_add_custom_inputs(\MoodleQuickForm &$mform) {
+        $mform->addElement('textarea', 'config_curl', get_string('connector_curl:curl', 'tool_dataflows'),
+                ['cols' => 50, 'rows' => 7]);
+        $mform->addElement('textarea', 'config_destination', get_string('connector_curl:destination', 'tool_dataflows'));
```
single line not textarea

brendanheywood replied:
also I cannot get this to work, every time I save a path eg '/tmp/postman.html'  then after I have saved it then it mangles it into destination: tmppostman.html

marcghaly replied:
I used wrong param type PARAM_FILE instead of PARAM_PATH, this is fixed.

brendanheywood commented on classes/local/step/connector_curl.php:
```diff
@@ -26,22 +26,183 @@
+        $errno = $curl->get_errno();
+        if ($httpcode >= 400 || empty($response) || $errno == 28) {
+            if (!empty($response)) {
+                $codestatustext = $this->get_error_status($httpcode);
```
Why do we need this lookup? We should use the text that we get from the response which might not be the same as this

marcghaly replied:
I have remove this - used text from response.

brendanheywood commented:
@marcghaly can you please rebase this, was a few clashes. Comments in the pr

marcghaly commented:
Hi @brendanheywood,

I have fixed issues above, path issue and `download_one` method issue which by default is not outputting proper response for `put` and `post`. It is rebased over latest master.

Regards,

Marc

marcghaly commented:
Hi @keevan,

This is a PR for issue #167, as you had that under radar is it possible for you to have a look at this ?

Thank you

Regards,

Marc

keevan commented:
@marcghaly Sure thing. I can take a look

brendanheywood commented on classes/local/step/connector_curl.php:
```diff
@@ -26,19 +26,166 @@
+        $mform->addElement('textarea', 'config_headers', get_string('connector_curl:headers', 'tool_dataflows'),
+                ['cols' => 50, 'rows' => 7]);
+        $mform->addElement('select', 'config_method', get_string('connector_curl:method', 'tool_dataflows'),
+                ['GET', 'POST', 'PUT']);
```
Can I suggest this is changed so the data saved is actually the method name and not an index number. Will make it much easier to debug and future proof when we add new methods like PATCH or reorder the UI here

marcghaly replied:
Fixed.

brendanheywood commented on lang/en/tool_dataflows.php:
```diff
@@ -231,3 +231,13 @@
+$string['connector_curl:headers'] = 'Headers';
+$string['connector_curl:method'] = 'HTTP request methods';
+$string['connector_curl:rawpostdata'] = 'Raw post data';
+$string['connector_curl:sideeffects'] = 'Has this request side effects ?';
```
This this out a little, eg

Does this request have side effects ?

Set this to true for API calls which change state on the remove server. This only affects whether the curl will be done or not in dry run mode or not.

keevan commented on classes/local/step/connector_curl.php:
```diff
@@ -26,19 +26,166 @@
+            }
+            if (isset($config->rawpostdata)) {
+                $options['CURLOPT_POSTFIELDS'] = $postdata;
+                if ($method == 'post') {
+                    $options['CURLOPT_POST'] = 1;
+                }
+                if ($method == 'put') {
+                    $options['CURLOPT_PUT'] = 1;
+                }
```
Do these options need to be nested under the postdata check? Are we able to reduce the amount of nesting here?

marcghaly replied:
Added them at the top as $options is used for every method. Fixed.

keevan commented on classes/local/step/connector_curl.php:
```diff
@@ -26,19 +26,166 @@
+        if (empty($config->curl)) {
+            $errors['config_curl'] = get_string('config_field_missing', 'tool_dataflows', 'curl', true);
+        }
+        if (($config->method == 1 || $config->method == 2) && empty($config->rawpostdata)) {
```
Can you please create class constants or similar to hold these different methods so they can be more easily identified rather than being magic numbers?

e.g. 
```
class ... {
    const METHOD_GET = 0;
    const METHOD_POST = 1;
    const METHOD_PUT = 2;
}
```

marcghaly replied:
Replaced those with method names as value are now direct names

keevan commented on classes/local/step/connector_curl.php:
```diff
@@ -26,19 +26,166 @@
+        $errno = $curl->get_errno();
+        if ($httpcode >= 400 || empty($response) || $errno == 28) {
+            if (is_string($result)) {
+                $resultarr = json_decode($result, true);
+                if (isset($resultarr['error'])) {
+                    $codestatustext = $httpcode . '/' . $resultarr['error'];
+                } else {
+                    $codestatustext = $httpcode . '/' . $result;
+                }
```
```suggestion
                $result = json_decode($result, true)['error'] ?? $result;
                $codestatustext = $httpcode . '/' . $result;
```

Would this work?

marcghaly replied:
Works better ! thank you

keevan reviewed (commented):
Hey @marcghaly I've done a pass of the code. I haven't tested the functionality of the code yet but I'll do that shortly.

keevan commented on classes/local/step/connector_curl.php:
```diff
@@ -26,19 +26,166 @@
+        $info = $curl->get_info();
+        // Stores response to be reusable by other steps.
+        $response = $curl->getResponse();
+        $this->enginestep->stepdef->config->response = $response;
```
Instead of setting the response here directly, could you use `$this->enginestep->set_var()` instead? (which was intended for similar purposes)

marcghaly replied:
Fixed.

keevan commented:
Doing a bit more functional testing (from the UI):

1. There is no mention of where the response is stored if a file destination is not set. This information could be useful if you're creating steps in order.
2. There should be some handling for dry mode, in particular dry mode shouldn't create any files even if the file path was set, and it shouldn't perform the cURL request if it's marked as having a side effect in configuration.

brendanheywood commented:
> There should be some handling for dry mode, in particular dry mode shouldn't create any files even if the file path was set, and it shouldn't perform the cURL request if it's marked as having a side effect in configuration.

To refine this a little, if curl downloads a file into the local scratch dir then this should NOT be considered having a side effect. It is not changing state outside of its own temporary memory and storages. If that file was then copied in another step to path which was outside the scratch dir then that would have side effects. So in theory the only time a curl step should have side effects is if the remote host is marked as changing state

@marcghaly this is a little tricky as the scratch directory pr hasn't landed yet:

https://github.com/catalyst/moodle-tool_dataflows/pull/252/files

Also if you are in dry run mode, and normally you are curling a remote api and it grabs a file which is then used in later steps, then our workflow will now abort in dry run mode when the later step can't find the file it needs. So what we need is an additional setting which is the contents of the file it will write in dry run mode when it will not actually do the real curl call. This can just be a textarea or a file upload which ever is easiest.

keevan commented:
I'm having trouble referencing the response from a subsequent step for when the destination path is not set. 

e.g. using the expression `steps.curl.config.response`

To ease with testing, are you able to export the dataflow you've been using to test with?

marcghaly commented on classes/local/step/connector_curl.php:
```diff
@@ -26,19 +26,166 @@
+            $this->hassideeffect = false;
+        }
+
+        if (isset($config->headers) && !empty($config->headers)) {
+            $headers = $config->headers;
+            if (!is_array($headers)) {
+                $headers = json_decode($headers, true);
+            }
+            $curl->setHeader($headers);
+        }
+        if (isset($config->rawpostdata)) {
+            $postdata = json_decode($config->rawpostdata, true);
+        }
```
Fixed.

marcghaly commented on classes/local/step/connector_curl.php:
```diff
@@ -26,19 +26,166 @@
+            $this->hassideeffect = false;
+        }
+
+        if (isset($config->headers) && !empty($config->headers)) {
```
Fixed.

marcghaly commented on classes/local/step/connector_curl.php:
```diff
@@ -26,19 +26,166 @@
+            }
+            $curl->setHeader($headers);
+        }
+        if (isset($config->rawpostdata)) {
```
If data is to be stored directly how do you handle multiple keys being posted ? for example :

{
    "name": "doe",
    "surname": "john",
    "id" : "57"
}

I have added a comment

brendanheywood replied:
Broadly the curl step should not make any assumptions about the data it is posting, or the data it is getting back. 

If the request post data happens to be json then it should be equivalent to this curl cli command:

```
curl -X POST https://reqbin.com/echo/post/json
   -H 'Content-Type: application/json'
   -d '{"login":"my_login","password":"my_password"}'
```

if it is a form post then it could be like this:

```
curl -X POST https://reqbin.com/echo/post/form
   -H "Content-Type: application/x-www-form-urlencoded" 
   -d "param1=value1&param2=value2" 
```

When you run it in dry run mode I would like the step to output a command like the above that can be cut and paste and run on the command line manually to make this dirt easy to debug

marcghaly replied:
Fixed, and added command

marcghaly commented on classes/local/step/connector_curl.php:
```diff
@@ -26,19 +26,166 @@
+        global $CFG, $PAGE;
+        // Get variables.
+        $config = $this->enginestep->stepdef->config;
+        $requesttype = [0 => 'get', 1 => 'post', 2 => 'put'];
```
I have removed this array as the select value is now the method name

marcghaly commented on classes/local/step/connector_curl.php:
```diff
@@ -26,19 +26,166 @@
      */
     public function execute(): bool {
-        // TODO: Implement.
+        global $CFG, $PAGE;
```
Fixed.

marcghaly commented on classes/local/step/connector_curl.php:
```diff
@@ -26,19 +26,166 @@
+                    $codestatustext = $httpcode . '/' . $result;
+                }
+            }
+            $this->enginestep->engine->abort(new \moodle_exception($codestatustext));
```
Fixed - just thrown exception

marcghaly commented on lang/en/tool_dataflows.php:
```diff
@@ -231,3 +231,13 @@
+$string['connector_curl:headers'] = 'Headers';
+$string['connector_curl:method'] = 'HTTP request methods';
+$string['connector_curl:rawpostdata'] = 'Raw post data';
+$string['connector_curl:sideeffects'] = 'Has this request side effects ?';
```
Fixed

marcghaly commented on classes/local/step/connector_curl.php:
```diff
@@ -26,19 +26,166 @@
+        }
+        // Perform call or download/call.
+        if (!empty($config->destination)) {
+            // Download.
```
Fixed.

marcghaly commented on classes/local/step/connector_curl.php:
```diff
@@ -26,19 +26,166 @@
+            if ($config->destination[0] === '/') {
+                $config->destination = substr($config->destination, 1);
+            }
+            if (isset($config->rawpostdata)) {
+                $options['CURLOPT_POSTFIELDS'] = $postdata;
+                if ($method == 'post') {
+                    $options['CURLOPT_POST'] = 1;
+                }
+                if ($method == 'put') {
+                    $options['CURLOPT_PUT'] = 1;
+                }
+            }
```
Fixed.

marcghaly commented on classes/local/step/connector_curl.php:
```diff
@@ -26,19 +26,166 @@
+        $this->enginestep->stepdef->config->response = $response;
+
+        $httpcode = $info['http_code'] ?? null;
+        $connecttime = $info['connect_time'] ?? null;
+        $totaltime = $info['total_time'] ?? null;
+        $sizeupload = $info['size_upload'] ?? null;
```
@brendanheywood mentioned in issue #167 that he wanted those stored variables - should I get rid of those or use set_vars instead ?

keevan replied:
Please set them via set_vars for now.

brendanheywood replied:
@marcghaly I'm ok to land this bit as is and we can clean this up in a followup issue.

This is what I would like, but the set_var api needs to be improved to allow it:

https://github.com/catalyst/moodle-tool_dataflows/issues/260

brendanheywood replied:
Sorry to clarify, I do eventually want all this timing data available, but wether it is sorted here or later I'm not too fussed. If this isn't working then comment it out and put a TODO for the later pr

keevan replied:
+1 adding a TODO comment above the block is fine for me.

marcghaly replied:
Adding a TODO

marcghaly commented:
Hi @keevan,

Thank you for the review, I made appropriate changes - are you able to use data from a former step this way `$this->enginestep->stepdef->get_variables()['steps']`, not sure if this `steps.curl.config.response` is yaml expression in which case I am not used to, do you have an example ?

Regards,

Marc

brendanheywood commented on classes/local/step/connector_curl.php:
```diff
@@ -26,19 +26,166 @@
+        // Perform call or download/call.
+        if (!empty($config->destination)) {
+            // Download.
+            if ($config->destination[0] === '/') {
+                $config->destination = substr($config->destination, 1);
+            }
```
Jason's patch hopefully will land today. If it does then lets use the scratch dir api, if it doesn't then lets land this PR regardless and make a follow up issue to convert curl to use the scratch dir

keevan replied:
Just a heads up, this PR has been merged so I'd use the new api now.

https://github.com/catalyst/moodle-tool_dataflows/pull/252

What you want is something like:
```
$this->enginestep->engine->resolve_path($destination);
```

marcghaly replied:
Fixed

brendanheywood commented on classes/local/step/connector_curl.php:
```diff
@@ -27,18 +27,176 @@
+                $options['CURLOPT_POSTFIELDS'] = $postdata;
+            }
+            $file = fopen($CFG->dirroot . '/' . $config->destination, 'w');
+            $result = $curl->$method($config->curl, $postdata, $options);
```
Why do we pull the downloaded file into memory (file could be huge) and then write it to a file? we can use CURLOPT_FILE to get curl to write it directly

brendanheywood replied:
After this is resolved I suspect this nested if's can be unravelled and we'll end up with only a single place which actually calls curl and the rest is just setting up options

marcghaly replied:
Used CURLOPT_FILE and removed nested if's

brendanheywood commented on classes/local/step/connector_curl.php:
```diff
@@ -27,18 +27,176 @@
+        if (($httpcode >= 400 || empty($response) || $errno == 28) && !$isdryrun) {
+            $codestatustext = null;
+            if (is_string($result)) {
+                $result = json_decode($result, true)['error'] ?? $result;
```
This appears to be expecting that the result from the curl will be json which is not true. Curl should not care in the slightest what the download is, it could be csv or binary

marcghaly replied:
Fixed.

marcghaly commented:
Hi @keevan,

I have fixed points cited above could you have another look if possible ?

Thank you 

Marc

keevan commented:
Hi @marcghaly , one other thing. I would like each type of step to have at least one unit test of some sort. Are you able to include one for this step please? I'll continue the review now.

keevan commented on classes/local/step/connector_curl.php:
```diff
@@ -27,18 +27,181 @@
+
+        $curl = new \curl();
+
+        if (!empty($config->sideeffects)) {
+            $this->hassideeffect = true;
+        } else {
+            $this->hassideeffect = false;
+        }
```
```suggestion
        $this->hassideeffect = !empty($config->sideeffects);
```

marcghaly replied:
Fixed

keevan commented on classes/local/step/connector_curl.php:
```diff
@@ -27,18 +27,181 @@
+        if (!$isdryrun) {
+            $result = $curl->$method($config->curl, [], $options);
+            // For put and post methods automatically overrides/has side effects.
+            if ($config->method != 'get') {
+                $this->hassideeffect = true;
+            }
```
What is the reason this field needs to be updated?

In either case, if it's needed I think the handling for this doesn't need to be here in particular, and can join the snippet further above where it's initially set. This should be deterministic before the curl is run, so should be further above before any actual handling.

marcghaly replied:
If method is `post` or `put` it will have side effects but setting it above will prevent the execution of the curl request

keevan replied:
Sorry I'm not following. I've tried moving the if block up and making it look like so, and it seems to work okay:
```
        $this->hassideeffect = !empty($config->sideeffects);
        if ($config->method != 'get') {
            $this->hassideeffect = true;
        }
```

The other question was whether or not $this->hassideeffect was used anywhere in this step in particular, since it's only used for display purposes and an example is located here https://github.com/catalyst/moodle-tool_dataflows/blob/MOODLE_35_STABLE/classes/local/step/flow_step.php#L109

I think if it can change, based on configuration values, then perhaps the has_side_effect function in base_step needs to be overriden in the connector_curl step, to include the extra checks. It doesn't seem like the block I mentioned above does anything in particular, but if you think it should could you please clarify?

keevan commented on classes/local/step/connector_curl.php:
```diff
@@ -27,18 +27,181 @@
+            }
+        }
+
+        if (!empty($config->destination) && !$isdryrun) {
+            fclose($file);
+        }
```
I think this can simply be 
```suggestion
        if (!empty($file)) {
            fclose($file);
        }
```

keevan reviewed (commented):
![image](https://user-images.githubusercontent.com/9924643/177451682-3f3a71d7-dd48-4094-9255-287784080415.png)

Also, can you please add some explainer/help text to the "headers" field I'm not sure what the format should be in. I assume it would be something like:
```
header-key: header value
another-key: 1234
...
etc.
```

marcghaly commented on classes/local/step/connector_curl.php:
```diff
@@ -27,18 +27,181 @@
      */
     public function execute(): bool {
-        // TODO: Implement.
+        global $CFG;
```
Fixed

marcghaly commented on classes/local/step/connector_curl.php:
```diff
@@ -27,18 +27,181 @@
+
+        // Download response to file provided destination is set.
+        if (!empty($config->destination) && !$isdryrun) {
+            if ($config->destination[0] === '/') {
+                $config->destination = substr($config->destination, 1);
+            }
```
used ltrm instead

marcghaly commented on classes/local/step/connector_curl.php:
```diff
@@ -27,18 +27,181 @@
+
+        $options = ['CURLOPT_TIMEOUT' => $this->timeout];
+
+        if ($method == 'post') {
```
fixed

marcghaly commented on classes/local/step/connector_curl.php:
```diff
@@ -27,18 +27,181 @@
+            $options['CURLOPT_POST'] = 1;
+        }
+
+        if ($method == 'put') {
```
fixed

marcghaly commented:
Hi @keevan, I have added unit tests - just unsure concerning the API I am using, should it be like a "mock api" so it is always "online", this is also rebased over last master.

Regards

Marc

keevan commented:
> Hi @keevan, I have added unit tests and fixed issues - just unsure concerning the API I am using, should it be like a "mock api" so it is always "online", this is also rebased over last master.
> 
> Regards
> 
> Marc

Hi Marc, I noticed there are various unaddressed comments on the files tab. Can you please respond to those open issues.

Also regarding the API test. It's much more reliable (and fast) if we mock the response instead of relying on a third party. There might be examples in core that do this already. From a high level, I'm expecting a combination of test things like:
- is the method set being called correctly?
- is the response being stored as a file correctly?
- is the response being stored into a variable correctly (where it can be referenced from future steps)
- shape of input/output data if any (more generally speaking)
- And anything else you'd want to check if someone were to ask, "How do you know this step works correctly?"

marcghaly commented:
Hi @keevan, I have updated unit tests let me know status concerning side effects issue

Regards,

Marc

keevan commented on lang/en/tool_dataflows.php:
```diff
@@ -251,3 +251,16 @@
+$string['connector_curl:sideeffects'] = 'Does this request have side effects?';
+$string['connector_curl:timeout'] = 'Timeout';
+$string['connector_curl:timeout_help'] = 'Time after which curl request will abort, default is 60 seconds.';
+$string['field_headers_help'] = 'Headers should be in the following JSON format: <br> {<br>"header-key": "header value", <br>"another-key": "1234"<br>}';
```
Could you please move the code example to the caller (since code snippets shouldn't be translated), and wrap them inside a pre tag to preserve formatting.

Currently it displays as follows:

![image](https://user-images.githubusercontent.com/9924643/178169984-1cc4298d-df29-451a-bd2b-8663898ffe32.png)

I'm expecting something like:

![image](https://user-images.githubusercontent.com/9924643/178170103-88188c26-1611-4e10-9832-013d388d09e0.png)

Also, have you considered the possibility of having YAML support for these headers? Since YAML is a superset of JSON, I think it should work with JSON, but yaml might be nicer display-wise

e.g.

![image](https://user-images.githubusercontent.com/9924643/178170197-c2a2a218-36b0-42c4-998c-227a44269f21.png)

(taken from https://www.json2yaml.com/ - ps: you can ignore line 1 of the yaml version)

marcghaly replied:
I have fixed display and move string to caller, also added yaml support and its corresponding example

keevan commented:
HI @marcghaly, could you please 
- update the PR description to have something like "Resolves #167" as this will ensure the linked issue is automatically closed off when this PR merges. Also it allows you to access this PR from the kanban board.
- resolve the version.php conflicts.
- Could you please add an `@author` tag to the curl connector file, as you've written the code and deserve the kudos
- there might be one extra point I've made prior to this comment, so if you could fix that up as well.

With regards the side effects setting of state. I think though it's not used currently, it might see some use later on, but it's quite harmless to leave in there. If anything, it would be moved to a `has_side_effects` function (overriding the base class), which will then affect how it displays in the UI, but that could be in a separate issue and addresssed when needed.

But for now, I'm quite happy with what it's currently demonstrating so if there are any bugs / issues once used with the json reader step then we'll iron those out then as well.

Thanks @marcghaly.

keevan commented on tests/tool_dataflows_curl_connector_test.php:
```diff
@@ -0,0 +1,210 @@
+use tool_dataflows\step;
+use tool_dataflows\dataflow;
+
+defined('MOODLE_INTERNAL') || die();
```
This is not required, so if you could please remove this line.

```

FILE: /admin/tool/dataflows/tests/tool_dataflows_curl_connector_test.php
--------------------------------------------------------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
--------------------------------------------------------------------------------------------------------------------
 26 | WARNING | [x] Unexpected MOODLE_INTERNAL check. No side effects or multiple artifacts detected.
--------------------------------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY
--------------------------------------------------------------------------------------------------------------------
```

marcghaly replied:
Fixed

marcghaly commented:
Hi @keevan, points above have been fixed - branch is rebased over latest main one.

Regards,

Marc

keevan reviewed (approved):
Looks good now, there'll be more adjustments made as we go along since there are other things available but this is good to be merged in.

keevan commented:
Test comment:

Resolves #167
